src/Microsoft.Diagnostics.Monitoring.EventPipe/Counters/TraceEventExtensions.cs

```diff
@@ -57,26 +58,269 @@ public static bool TryGetCounterPayload(this TraceEvent traceEvent, CounterFilte
+                {
+                    HandleObservableInstrumentCallbackError(traceEvent, sessionId, out individualPayload);
+                }
+                else if (traceEvent.EventName == "MultipleSessionsNotSupportedError")
+                {
+                    HandleMultipleSessionsNotSupportedError(traceEvent, sessionId, out individualPayload);
+                }
+
+                if (null != individualPayload)
+                {
+                    payload.Add(individualPayload);
+                }
+
+                return null != payload && payload.Any();
+            }
+
             return false;
         }
 
+        public static bool TryGetIndividualCounterPayload(this TraceEvent traceEvent, CounterFilter filter, out ICounterPayload payload)
+        {
+            bool gotCounterPayload = TryGetCounterPayload(traceEvent, filter, null, out List<ICounterPayload> payloadsList);
+
+            payload = payloadsList.FirstOrDefault();
+
+            return gotCounterPayload;
+        }
+
+        private static void HandleGauge(TraceEvent obj, CounterFilter filter, string sessionId, out ICounterPayload payload)
+        {
+            payload = null;
+
+            string payloadSessionId = (string)obj.PayloadValue(0);
+
+            if (payloadSessionId != sessionId)
+            {
+                return;
+            }
+
+            string meterName = (string)obj.PayloadValue(1);
+            //string meterVersion = (string)obj.PayloadValue(2);
+            string instrumentName = (string)obj.PayloadValue(3);
+            string unit = (string)obj.PayloadValue(4);
+            string tags = (string)obj.PayloadValue(5);
+            string lastValueText = (string)obj.PayloadValue(6);
+
+            if (!filter.IsIncluded(meterName, instrumentName))
+            {
+                return;
+            }
+
+            // the value might be an empty string indicating no measurement was provided this collection interval
+            if (double.TryParse(lastValueText, out double lastValue))
+            {
+                payload = new GaugePayload(meterName, instrumentName, null, unit, tags, lastValue, obj.TimeStamp);
+            }
+            else
+            {
+                // for observable instruments we assume the lack of data is meaningful and remove it from the UI
+                // this happens when the Gauge callback function throws an exception.
+
+                //TODO Can this occur for other meter types?
+                payload = new CounterEndedPayload(meterName, instrumentName, null, obj.TimeStamp);
+                
+            }
+        }
+
+        private static void HandleCounterRate(TraceEvent traceEvent, CounterFilter filter, string sessionId, out ICounterPayload payload)
+        {
+            payload = null;
+
+            string payloadSessionId = (string)traceEvent.PayloadValue(0);
+
+            if (payloadSessionId != sessionId)
+            {
+                return;
+            }
```

Comment:
We need to add the `if (!filter.IsIncluded(meterName, instrumentName))` check for this one - it was inadvertently omitted in my initial draft.